R/quantile-pred.R

```diff
@@ -0,0 +1,219 @@
+    cli::cli_abort(
+      "{.arg quantile_levels} must be sorted in increasing order.",
+      call = call
+    )
+  }
+  invisible(NULL)
+}
+
+check_vector_probability <- function(x, ...,
+                                     allow_na = FALSE,
+                                     allow_null = FALSE,
+                                     arg = caller_arg(x),
+                                     call = caller_env()) {
+  for (d in x) {
+    check_number_decimal(
+      d, 
+      min = 0,
+      max = 1,
+      arg = arg,
+      call = call,
+      allow_na = allow_na,
+      allow_null = allow_null,
+      allow_infinite = FALSE
+    )
+  }
+}
+
+check_quantile_level <- function(x, object, call) {
+  if (object$mode != "quantile regression") {
+    return(invisible(TRUE))
+  } else {
+    if (is.null(x)) {
+      cli::cli_abort("In {.fn check_mode}, at least one value of
+      {.arg quantile_level} must be specified for quantile regression models.")
+    }
+  }
+  if (any(is.na(x))) {
+    cli::cli_abort("Missing values are not allowed in {.arg quantile_levels}.",
+                   call = call)
+  }
+  x <- sort(unique(x))
+  check_vector_probability(x, arg = "quantile_level", call = call)
+  x
+}
+
+#' @export
+#' @rdname quantile_pred
+extract_quantile_levels <- function(x) {
+  if (!inherits(x, "quantile_pred")) {
+    cli::cli_abort("{.arg x} should have class {.cls quantile_pred}.")
+  }
+  attr(x, "quantile_levels")
+}
+
+#' @export
+#' @rdname quantile_pred
+as_tibble.quantile_pred <-
+  function (x, ..., .rows = NULL, .name_repair = "minimal", rownames = NULL) {
+    lvls <- attr(x, "quantile_levels")
+    n_samp <- length(x)
+    n_quant <- length(lvls)
+    tibble::new_tibble(list(
+      .pred_quantile = unlist(x),
+      .quantile_levels = rep(lvls, n_samp),
+      .row = rep(1:n_samp, each = n_quant)
+    ))
+  }
+
+#' @export
+#' @rdname quantile_pred
+as.matrix.quantile_pred <- function(x, ...) {
+  num_samp <- length(x)
+  matrix(unlist(x), nrow = num_samp)
+}
+
+#' @export
+format.quantile_pred <- function(x, ...) {
+  quantile_levels <- attr(x, "quantile_levels")
+  if (length(quantile_levels) == 1L) {
+    x <- unlist(x)
+    out <- round(x, 3L)
+    out[is.na(x)] <- NA_real_
+  } else {
+    rng <- sapply(x, range, na.rm = TRUE)
+    out <- paste0("[", round(rng[1, ], 3L), ", ", round(rng[2, ], 3L), "]")
+    out[is.na(rng[1, ]) & is.na(rng[2, ])] <- NA_character_
+    m <- median(x)
+    out <- paste0("[", round(m, 3L), "]")
+  }
+  out
+}
+
+#' @export
+median.quantile_pred <- function(x, ...) {
+  lvls <- attr(x, "quantile_levels")
+  loc_median <- (abs(lvls - 0.5) < sqrt(.Machine$double.eps))
+  if (any(loc_median)) {
+    return(map_dbl(x, ~ .x[min(which(loc_median))]))
+  }
+  if (length(lvls) < 2 || min(lvls) > 0.5 || max(lvls) < 0.5) {
+    return(rep(NA, vctrs::vec_size(x)))
+  }
+  map_dbl(x, ~ stats::approx(lvls, .x, xout = 0.5)$y)
+}
+
+#' @importFrom vctrs vec_ptype_abbr
+#' @export
```

Comment:
I believe that this is resolved by [d0444bd](https://github.com/tidymodels/hardhat/pull/260/commits/d0444bdae0fbc42b9f1b9a21f2d34aaa9c7710d8) but please verify

Reply:
I've removed the re-export in [c597214](https://github.com/tidymodels/hardhat/pull/260/commits/c5972148dd07a42f9e71fc0c05f6365f1e22e441). If my intuition here was wrong and we do need them exported, we can do so again. I'd just like us to bump into this first instead of preemptively exporting them.